File: #1.data-structures/#2.Stack/src/c#/src/MyStackProject/modules/UndoRedo.cs
public class UndoRedo<T>
{
  private Stack<T> undoStack = new Stack<T>();
  private Stack<T> redoStack = new Stack<T>();

  public void AddAction(T action) 
  {
    undoStack.Push(action);
    redoStack = new Stack<T>();
  }

  public T? Undo()
  {
    var action = undoStack.Pop();
    if(action != null) {
      redoStack.Push(action);
    }

    return action;
  }

  public T? Redo()
  {
    var action = redoStack.Pop();
    if(action != null) {
      undoStack.Push(action);
    }
    return action;
  }

  public void Debug() {
    Console.WriteLine("Undo Stack: " + string.Join(',', undoStack.GetItems()));
    Console.WriteLine("Redo Stack: " + string.Join(',', redoStack.GetItems()));
  }

}

File: #1.data-structures/#2.Stack/src/swfit/Sources/modules/UndoRedo.swift
class UndoRedo<T> {
    private let undoStack = Stack<T>()
    private var redoStack = Stack<T>()

    func addAction(_ action: T) {
        undoStack.push(action)
        redoStack = Stack() // Limpa a pilha de redo
    }

    func undo() -> T? {
        guard let action = undoStack.pop() else {
            return nil
        }
        redoStack.push(action)
        return action
    }

    func redo() -> T? {
        guard let action = redoStack.pop() else {
            return nil
        }
        undoStack.push(action)
        return action
    }

    func debug() {
        print("Undo Stack: \(undoStack.getItems())")
        print("Redo Stack: \(redoStack.getItems())")
    }
}
